src/utils/triton.h

```diff
@@ -0,0 +1,180 @@
+    return static_cast<dali_data_type_t>(0);
+  return static_cast<dali_data_type_t>(t - 2);
+}
+
+/**
+ * @brief Converts TRITONSERVER_MemoryType to DALI device_type_t
+ */
+inline device_type_t to_dali(TRITONSERVER_MemoryType t) {
+  switch (t) {
+    case TRITONSERVER_MEMORY_CPU:
+    case TRITONSERVER_MEMORY_CPU_PINNED:
+      return device_type_t::CPU;
+    case TRITONSERVER_MEMORY_GPU:
+      return device_type_t::GPU;
+    default:
+      throw std::invalid_argument("Unknown memory type");
+  }
+}
+
+/**
+ * @brief Converts dali_data_type_t to TRITONSERVER_DataType
+ */
+inline TRITONSERVER_DataType to_triton(dali_data_type_t t) {
+  assert(-1 <= t && t <= 11);
+  // Use the trick, that TRITONSERVER_DataType and dali_data_type_t
+  // types have more-or-less the same order, with few exceptions
+  if (t == -1)
+    return static_cast<TRITONSERVER_DataType>(0);
+  if (t == 11)
+    return static_cast<TRITONSERVER_DataType>(1);
+  return static_cast<TRITONSERVER_DataType>(t + 2);
+}
+
+/**
+ * @brief Converts device_type_t to TRITONSERVER_MemoryType
+ */
+inline TRITONSERVER_MemoryType to_triton(device_type_t dev) {
+  switch (dev) {
+    case CPU:
+      return TRITONSERVER_MEMORY_CPU_PINNED;
+    case GPU:
+      return TRITONSERVER_MEMORY_GPU;
+    default:
+      throw std::invalid_argument("Unknown memory type");
+  }
+}
+
+class TritonInput {
+ public:
+  TritonInput(TRITONBACKEND_Input *handle) : handle_(handle) {
+    TRITONSERVER_DataType input_datatype;
+    const int64_t *input_shape;
+    uint32_t input_dims_count;
+    uint64_t input_byte_size;
+    uint32_t input_buffer_count;
+    const char *name;
+    TRITON_CALL_GUARD(TRITONBACKEND_InputProperties(handle_, &name, &input_datatype, &input_shape,
+                                                    &input_dims_count, &byte_size_, &buffer_cnt_));
+    meta_.name = std::string(name);
+    meta_.type = to_dali(input_datatype);
+    auto batch_size = input_shape[0];
+    auto sample_shape = TensorShape<>(input_shape + 1, input_shape + input_dims_count);
+    auto shape = TensorListShape<>::make_uniform(batch_size, sample_shape);
+    meta_.shape = shape;
+  }
+
+  IOMeta Meta() const {
+    return meta_;
+  }
+
+  size_t ByteSize() const {
+    return byte_size_;
+  }
+
+  uint32_t BufferCount() const {
+    return buffer_cnt_;
+  }
+
+  /**
+   * @brief Request an input buffer.
+   * @param idx Input index.
+   * @param device_type_t Preferred device type.
+   * @param device_id Preferred device id.
+   */
+  IBufferDescr GetBuffer(uint32_t idx, device_type_t device, int device_id) {
+    const void *data;
+    size_t size;
+    TRITONSERVER_MemoryType mem_type = to_triton(device);
+    int64_t mem_type_id = device_id;
+    TRITON_CALL_GUARD(
+        TRITONBACKEND_InputBuffer(handle_, idx, &data, &size, &mem_type, &mem_type_id));
+    IBufferDescr descr;
+    descr.device = to_dali(mem_type);
+    descr.device_id = mem_type_id;
+    descr.data = data;
+    descr.size = size;
```

Comment:
Add initializers to `BufferDescr` - otherwise this code is not future proof and a potential correctness or even security issue.